src/solver/README.md

```diff
@@ -0,0 +1,238 @@
+
+#### Success case
+
+If the test returns a success, the extensions of the candidate is generated:
+```text
+      ▼
+┌┲━┱┬┲━┱┬───┐
+│┃0┃│┃0┃│   │
+└╂─╂┴╂─╂┴───┘
+ ┃ ┃ ┃ ┃ 
+ ┗━┛ ┗━┛
+```
+
+When a candidate is complete, in other words, when the last cell of a candidate has been tested and did not lead to a backtrack, we know that the candidate is a solution to the problem, and we can yield it from the iterator.
+
+#### Failure case
+
+If the test returns a failure, the next step is to evaluate alternative source for the same resource. Let's assume that *Source 0* had *Resource A* but it does not have *Resource B*. In such case, the algorithm will increment the cell's index:
+
+```text
+      ▼
+     ┏━┓
+     ┃0┃
+┌┲━┱┬╂─╂┬───┐
+│┃0┃│┃1┃│   │
+└╂─╂┴┺━┹┴───┘
+ ┃ ┃    
+ ┗━┛ 
+ ```
+
+and that will potentially lead to a partial solution `[0, 1, ]` to be stored for the next iteration.
+
+If the test fails and no more sources can be generated, the algorithm will *backtrack* from to previous sources looking for a cell with the **highest** index prior to the cell that was being evaluated which is not yet on the last source. If such cell is found, the results of all cells **to the right** of the newfound cell are **erased** and the next branch can be evaluated.
+
+If no such cell can be found, that means that the iterator is complete.
+
+### Parallel
+
+If the testing can be performed in parallel, like an asynchronous I/O, the above *serial* solution is sub-optimal as it misses on the benefit of testing multiple cells at once.
+
+In such scenario, a difference to the algorithm is that it will not first populate a complete candidate such that all cells are at indexes that may lead to a solution and then call the testing method on such complete candidate:
+
+```text
+          ▼
+┌┲━┱┬┲━┱┬┲━┱┐
+│┃0┃│┃0┃│┃0┃│
+└╂─╂┴╂─╂┴╂─╂┘
+ ┃ ┃ ┃ ┃ ┃ ┃
+ ┗━┛ ┗━┛ ┗━┛
+```
+
+The result of the testing function in such case, is a set of results for each cell allowing the algorithm to accept or reject the candidate, and memoize which branches should be skipped in the later evaluations.
+
+#### Success case
+
+If the result a set of successes, the candidate is returned as a solution, and the algorithm proceeds to the same operation as if it was a failure.
+
+#### Failure case
+If the result contains failures, the iterator will now backtrack to find the closest lower or equal cell to the current index which can be advanced to the next source.
+In the example state above, the current cell can be advanced to *source 1* and then just a set of `[None, None, 1]` is to be evaluated by the tester (since we know that *A0* and *B0* are valid).
+
+If that is successful, the `[0, 0, 1]` set is a complete solution and is yielded.
+
+Then, the next state to be tested is:
```

Comment:
If we've found a complete solution, does the algorithm continue iterating? Or is this the failure case for the above?

Reply:
It yields, and if it is resumed later, it will try to build the next complete solution.

Reply:
Ok, that makes sense. Maybe make this say "Then, if the iterator is resumed, the next state..."